spotlight-client/src/datastores/Metric.ts

```diff
@@ -0,0 +1,316 @@
+  demographicView?: RecordFormat extends DemographicFields
+    ? DemographicView
+    : undefined;
+
+  constructor({
+    contentSource,
+    dataTransformer,
+    defaultLocalityId,
+    defaultDemographicView,
+    sourceFileName,
+  }: InitParams<RecordFormat>) {
+    // initialize metadata
+    this.description = contentSource.description;
+    this.methodology = contentSource.methodology;
+    this.name = contentSource.name;
+
+    // initialize data fetching
+    this.isLoading = true;
+    this.dataTransformer = dataTransformer;
+    this.sourceFileName = sourceFileName;
+
+    // initialize filters
+    this.localityId = defaultLocalityId;
+    this.demographicView = defaultDemographicView;
+  }
+
+  async fetch(): Promise<void> {
+    // TODO: map metric type to file name(s) in factory function?
+    const apiResponse = await callMetricsApi({
+      files: [this.sourceFileName],
+    });
+    // TODO: call the proper data transformation function for the metric type
+    if (apiResponse) {
+      const metricFileData = apiResponse[this.sourceFileName];
+      if (metricFileData) {
+        this.allRecords = this.dataTransformer(metricFileData);
+      }
+      this.isLoading = false;
+    }
+  }
+
+  get records(): RecordFormat[] {
+    let recordsToReturn = this.allRecords || [];
+
+    if (this.localityId) {
+      recordsToReturn = recordsToReturn.filter(
+        (record) => record.locality === this.localityId
+      );
+    }
+
+    if (this.demographicView) {
+      recordsToReturn = recordsToReturn.filter(
+        recordIsTotalByDimension(this.demographicView)
+      );
+    }
+
+    return recordsToReturn;
+  }
+}
+
+export default Metric;
+
+/**
+ * Returns a filter predicate for the specified demographic view
+ * that will exclude totals and breakdowns for all other views
+ */
+function recordIsTotalByDimension(
+  demographicView: Exclude<DemographicView, "nofilter">
+): (record: RawRecord) => boolean {
+  const keysEnum = { ...DIMENSION_DATA_KEYS };
+
+  if (demographicView !== "total") {
+    delete keysEnum[demographicView];
+  }
+
+  const otherDataKeys = Object.values(keysEnum);
+
+  return (record) => {
+    let match = true;
+    if (demographicView !== "total") {
+      // filter out totals
+      match =
+        match && record[DIMENSION_DATA_KEYS[demographicView]] !== TOTAL_KEY;
+    }
+
+    // filter out subset permutations
+    match = match && otherDataKeys.every((key) => record[key] === TOTAL_KEY);
+
+    return match;
+  };
+}
+
+export type MetricsMapping = {
+  [MetricTypes.SentencePopulationCurrent]?: PopulationSnapshot;
+  [MetricTypes.SentenceTypesCurrent]?: SentenceTypesCurrent;
+};
+
+function extractBaseFields(record: ValuesType<DataFile>): BaseMetricFields {
+  return {
+    stateCode: record.state_code,
+  };
+}
+
+function extractDemographicFields(
+  record: ValuesType<DataFile>
+): DemographicFields {
+  return {
+    // we are trusting the API to return valid strings here, not validating them
```

Comment:
I've been discussing with the Lantern team about how we can strengthen this trust a bit more. We have been discussing adding tests and validations on the dashboard views to ensure that the values do not change under our feet, and that there is a process in place to change values without breaking the FE. We have an epic open for that, hopefully that eases your mind here a bit.

Reply:
yeah that would be great! Much better to do that once per pipeline run than have every client do it for every fetch